### PR TITLE
Support for shareless members

### LINKED ISSUE #1021: Support for share-less members
#51 introduced recovery shares for each member. However, there may be cases where only _some_ members are privileged enough to be able to recover a crashed service. In this case, only these members should be handed recovery shares. 

Notes:
- When adding a new member, only members that have a public encryption key should be handed recovery shares. Other members should not be counted in the total number of shares generated.
- It should be possible to determine from the Lua constitution whether a member has been given a recoverable key share. 
- If a share-less member is added or removed, there is no need for CCF to issue new share, not re-key the ledger.

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves #1021 

~(Opening as draft are there are still a couple of TODOs)~

This PR adds support for share-less members, i.e. members without a recovery share. Other members (i.e. those with recovery shares) are named "recovery members" in the documentation.

If a member is registered in CCF without a public encryption key, such member is not given a recovery share and thus cannot request/decrypt a recovery share. 

Changes:
- `cchost`'s `--member-info` now doesn't require the path of the member's public encryption key. In other words, `--member-info member_cert.pem`, `--member-info member_cert.pem,,member_data.json` (not the double `,,`) are valid.
- Recovery threshold, adding/removing members logic updated accordingly, so that the CCF service never ends up in a non-recoverable state (i.e. number of active recovery members < recovery threshold).
- Python infra and documentation

I have moved the recovery share end-to-end test code from `memberclient.py` to a new `membership.py` file.

### CODE REVIEW

File: doc/quickstart/test_network.rst [hunk 1]
```diff
@@ -1,21 +1,20 @@
...
-The script creates a new one node test CCF network running locally. All the governance requests required to open the network to users are automatically issued.
+The script creates a new one node CCF test network running locally. All the governance requests required to open the network to users are automatically issued.
 
 For example, deploying the ``liblogging`` example application:
 
 .. code-block:: bash
 
-    $ cd CCF/build
-    $ ../sandbox.sh --package ./liblogging.virtual.so
+    $ /opt/ccf/bin/sandbox.sh --package ./liblogging.virtual.so
```
Comment (jumaffre) [inline doc/quickstart/test_network.rst:17]: The paths seemed wrong when using the sandbox from an install of CCF.

File: python/ccf/proposal_generator.py [hunk 2]
```diff
@@ -231,11 +237,6 @@ def new_member(
     return false
     end
 
-    expected_enc_pub_key = [====[{encryption_pub_key}]====]
```
Comment (jumaffre) [inline python/ccf/proposal_generator.py:234]: This is now optional so the sample vote should only check for the member's identity public key.

File: src/host/main.cpp [hunk 3]
```diff
@@ -453,20 +453,36 @@ int main(int argc, char** argv)
 
     if (*start)
     {
+      // Count members with public encryption key as only these members will be
+      // handed a recovery share.
+      // Note that it is acceptable to start a network without any member having
+      // a recovery share. The service will check that at least one recovery
+      // member is added before the service can be opened.
```
Comment (jumaffre) [inline src/host/main.cpp:460]: Tested in `membership.py`, starting a CCF service with different combinations of members.

File: doc/operators/start_network.rst [hunk 4]
```diff
@@ -35,7 +35,7 @@ When starting up, the node generates its own key pair and outputs the certificat
...
 
-The certificates and recovery public keys of initial members of the consortium are specified via ``--member-info``. For example, if 3 members should be added to CCF, operators should specify ``--member-info member1_cert.pem,member1_enc_pubk.pem``, ``--member-info member2_cert.pem,member2_enc_pubk.pem``, ``--member-info member3_cert.pem,member3_enc_pubk.pem``.
+The certificates, encryption public keys and member data of initial members of the consortium are specified via ``--member-info``. For example, if 3 members (one recovery member with member data, one recovery member without member data and one non-recovery member with member data) should initially be added to CCF, operators should specify ``--member-info member1_cert.pem,member1_enc_pubk.pem,member1_data.json``, ``--member-info member2_cert.pem,member2_enc_pubk.pem``, ``--member-info member3_cert.pem,,member2_data.json`` (note the empty public encryption key path "``,,``" for the third member).
```
Comment (eddyashton) [inline doc/operators/start_network.rst:38]: I think this example is now covering too many things, and they should be separately listed. Something like:

> For example:
> - A recovery member with member data
> `--member-info member1_cert.pem,member1_enc_pubk.pem,member1_data.json`
> - A recovery member without member data
> `--member-info member2_cert.pem,member2_enc_pubk.pem`
> - A non-recovery member with member data (note the empty public encryption key path, indicating a non-recovery member)
> `--member-info member3_cert.pem,,member2_data.json`

File: python/ccf/proposal_generator.py (hunk 2 above)
Comment (eddyashton) [reply python/ccf/proposal_generator.py:234]: I think this should remain, wrapped in `if encryption_pub_key is not None:`. If the proposal _does_ specify an encryption public key, then the careful vote for it should check that encryption public key is present.

Comment (eddyashton) [reply python/ccf/proposal_generator.py:234]: Sorry this is inline Lua. So either formatted or hand-appended or something, if we have an encryption key in the Python we should check it in the Lua.

File: src/node/genesis_gen.h [hunk 5]
```diff
@@ -297,12 +315,12 @@ namespace ccf
     {
       auto service_view = tx.get_view(tables.service);
 
-      if (get_active_members_count() < get_recovery_threshold())
+      if (get_active_recovery_members().size() < get_recovery_threshold())
```
Comment (eddyashton) [inline src/node/genesis_gen.h:318]: Nit: `get_active_recovery_members()` is expensive, so would prefer:
```
const auto recovery_members_count = get_active_recovery_members().size();
if (recovery_members_count...)
{
  ..., recovery_members_count, ...
}
```

File: python/ccf/proposal_generator.py (hunk 2 above)
Comment (jumaffre) [reply python/ccf/proposal_generator.py:234]: You're absolutely right. I'd assumed that checking for the member's identity certificate was sufficient for some reason anyway. Fixed now.

File: doc/operators/start_network.rst (hunk 4 above)
Comment (jumaffre) [reply doc/operators/start_network.rst:38]: Good point. I've also added a bullet point for non-recovery, no member data member, i.e. `--member-info member_cert.pem`.

File: src/node/genesis_gen.h (hunk 5 above)
Comment (jumaffre) [reply src/node/genesis_gen.h:318]: Good spot.

File: src/node/rpc/member_frontend.h [hunk 6]
```diff
@@ -1733,7 +1747,16 @@ namespace ccf
           g.add_member(info);
         }
 
-        g.set_recovery_threshold(in.recovery_threshold);
+        // Note that it is acceptable to start a network without any member
+        // having a recovery share. The service will check that at least one
+        // member with a recovery is added before the service is opened.
```
Comment (achamayou) [inline src/node/rpc/member_frontend.h:1752]:
```suggestion
        // member with a recovery share is added before the service is opened.
```

**Comment by achamayou** [comment]
I think it would be good to add a check in the operator governance, validating that operators (ie. members with member_info.is_operator) do no have a recovery share, since it is the primary use case for that feature.

**Review by achamayou — APPROVED** [review]

**Comment by jumaffre** [comment]
> I think it would be good to add a check in the operator governance, validating that operators (ie. members with member_info.is_operator) do no have a recovery share, since it is the primary use case for that feature.

@achamayou: Done now 👍

**Pull request merged**